content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+    }
+}
+
+class AnotherClassToBreakImmutableObject extends SomeImmutableObject
+{
+
+    public function __construct(string $value)
+    {
+        $this->someString = $value;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+
+    public function change(): void
+    {
+        $this->someString .= ' and Minny';
+    }
+}
+
+$three = new AnotherClassToBreakImmutableObject('Pippo');
+echo $three->getValue(); //Pippo
+echo $three->change();
+echo $three->getValue(); //the value is: Pippo and Minny -
+```
+try it! **[break the immutable object] (https://3v4l.org/KpnfR)**
+
+It still does not work, ok now we definitely correct promised!
+The problem is that inheritance breaks the encapsulation so a good solution in this case is to use **final**!.
+**the definitive version**
+
+# The concrete example definitive fix with final keyword
+
+```php
+<?php
+
+final class SomeImmutableObject
+{
+    private $someString;
+    private $flagCreate = false;
+
+    public function __construct(string $value)
+    {
+
+        if ($this->flagCreate === true) {
+            throw new \BadMethodCallException('This is an immutable object has already create.');
+        }
+
+        $this->someString = $value;
+        $this->flagCreate = true;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+}
+
+class TryToBreakImmutableObject extends SomeImmutableObject
+{
+    public $someString;
+
+    public function __construct(string $value)
+    {
+        $this->someString = $value;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+
+    public function change(): void
+    {
+        $this->someString .= ' and Minny';
+    }
+}
+
+$four = new TryToBreakImmutableObject('Pippo');
+echo $four->getValue(); //Pippo
+echo $four->change();
+echo $four->getValue();
+```
+
+```PHP Fatal error:  Class BreakImmutableObject may not inherit from final class (SomeImmutableObject)```
+
+try it! **[immutable object] (https://3v4l.org/OsOoW)**
+
+
+… Some links in-depth study
+
+- [Final or not final] (https://github.com/thephpleague/period/issues/54) 
+- [Final classes by default] (https://github.com/symfony/symfony/issues/15233)
+- [Final classes: Open for Extension, Closed for Inheritance] (http://verraes.net/2014/05/final-classes-in-php/)
+
+
+Obviously it is not always all pink and flowers, not always literature especially in our work, it is perfectly suited to all needs. 
+What I want to point out is that in some cases **final** can be a good ally and can help us to simplify the design.  
+Using **final** keyword leads programmers to make a round of additional reasoning, to pay more attention to the use of the class. 
+For example, during review a diff with the removal of **final** keyword, it could lead to useful comments to find different solutions. 
```

Comment:
For example, during review, a diff with the removal of **final** keyword could lead to useful comments to find different solutions.